File: 14-TypeScript/Grey Multi-Tenant/packages/grey-ruby/lib/grey_sdk/domain/user_client.rb
# frozen_string_literal: true

module GreySdk
  module Domain
    # User client for user-related operations
    class UserClient
      # @param http_client [GreySdk::Http::HttpClient] The HTTP client
      def initialize(http_client)
        @http = http_client
      end

      # Retrieves a user by ID
      # @param user_id [String] The user ID
      # @return [GreySdk::Error::Result] Result containing user data or error
      def get_user(user_id:)
        return validation_error("User ID is required") if user_id.nil? || user_id.to_s.empty?

        @http.get("/api/v1/users/#{user_id}")
      end

      # Retrieves the current authenticated user
      # @return [GreySdk::Error::Result] Result containing user data or error
      def get_current_user
        @http.get("/api/v1/users/me")
      end

      # Updates a user's profile
      # @param user_id [String] The user ID
      # @param data [Hash] User data to update
      # @return [GreySdk::Error::Result] Result containing updated user or error
      def update_user(user_id:, data:)
        return validation_error("User ID is required") if user_id.nil? || user_id.to_s.empty?
        return validation_error("Update data is required") if data.nil? || data.empty?

        @http.patch("/api/v1/users/#{user_id}", data)
      end

      private

      def validation_error(message)
        Error::Result.err(Error::GreyError.validation(message))
      end
    end
  end
end

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-r/R/user_client.R
#' @title User Client
#' @description User client for user-related operations
#' @name UserClient
NULL

#' @importFrom R6 R6Class
#' @export
UserClient <- R6::R6Class(
  "UserClient",
  
  public = list(
    #' @description Create a new user client
    #' @param http_client HttpClient instance
    initialize = function(http_client) {
      private$http <- http_client
    },
    
    #' @description Get a user by ID
    #' @param user_id The user ID
    #' @return grey_result with user data or error
    get_user = function(user_id) {
      if (is.null(user_id) || nchar(trimws(as.character(user_id))) == 0) {
        return(result_err(error_validation("User ID is required")))
      }
      
      private$http$get(paste0("/api/v1/users/", user_id))
    },
    
    #' @description Get the current authenticated user
    #' @return grey_result with user data or error
    get_current_user = function() {
      private$http$get("/api/v1/users/me")
    },
    
    #' @description Update a user's profile
    #' @param user_id The user ID
    #' @param data Update data (list)
    #' @return grey_result with updated user or error
    update_user = function(user_id, data) {
      if (is.null(user_id) || nchar(trimws(as.character(user_id))) == 0) {
        return(result_err(error_validation("User ID is required")))
      }
      
      if (is.null(data) || length(data) == 0) {
        return(result_err(error_validation("Update data is required")))
      }
      
      private$http$patch(paste0("/api/v1/users/", user_id), data)
    }
  ),
  
  private = list(
    http = NULL
  )
)
